Fix thread safety problem in PhysicsTools/MVAComputer

Dr15Jones opened the pull request:
The CMSSW_7_4_THREADED_X IBs have been crashing due to MVA threading issues. The problem stemmed from the use of mutable in ProcForeach. These mutables were being read/written indirectly by GenericMVAJetTagComputer which is an EventSetup product. The simultaneous writes were causing the crashes.
The solution was to move the mutables to be external to the MVA class and instead be passed in as an argument to the functions. Encapsolating the loop state into variables that live on the stack allows each thread to have its own copy.

Dr15Jones committed:
Make ProcForeach used by MVAComputer thread safe

ProcForeach was using mutable member data to keep track of the
state of the looping it was doing. Given that multiple threads
could be updating those values concurrently, this lead to crashes.
The code has now been changed to store the looping state externally
via VarProcessor::LoopCtx on the stack inside MVAComputer::evalInternal.
This allows each thread to have its own copy.

Dr15Jones committed:
Minor cleanup of AtomicId

Theoretically, making the allocator static could lead to an order
of initialization problem. In reality the default allocator is
stateless so wasn't a problem. However not being static simplifies
understanding the code.
Since non of the member functions are const we don't need to use
mutable.

cmsbuild commented:
A new Pull Request was created by @Dr15Jones (Chris Jones) for CMSSW_7_4_X.

Fix thread safety problem in PhysicsTools/MVAComputer

It involves the following packages:

PhysicsTools/MVAComputer


@cmsbuild, @vadler, @nclopezo, @monttj can you please review it and eventually sign? Thanks.
@ferencek, @acaudron, @pvmulder, @imarches this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.

Dr15Jones commented:
@slava This problem was affecting RECO so you might wish to run your own tests.

Dr15Jones committed:
Make ProcessRegistry usage thread-safe

The static analyzer found that the registries were being accessed
at event processing time by some modules. Given that we could get
a shared library load happening while some other code was reading
the registry, we needed to make the lookup and registration process
threads safe. This was easy to accomplish using tbb::concurrent_unordered_map.

Dr15Jones commented:
I took the opportunity to also fix a threading problem related to registration found by the static analyzer. I have no further changes I plan on making.

cmsbuild commented:
Pull request #7332 was updated. @cmsbuild, @vadler, @nclopezo, @monttj can you please check and sign again.

Slava commented:
@Dr15Jones you are tagging the wrong person

Dr15Jones commented:
@slava77 You might want to check this out since the problem was originally see in RECO.

slava77 commented:
@cmsbuild please test

degano commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 365ba0a3ef9fe3133d5b4164adcb0daa79588f94
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7332/2064/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7332/2064/summary.html

slava77 commented:
hold

Hi Chris, 
remind me, please, was the issue in reco with just a crash in MT mode or with changes in performance as well.
I see that jenkins tests show differences in ghostTrack variables
e.g., this one is quite substantial 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_4_X_2015-01-23-0200+7332/5677/validateJR/all_OldVSNew_RunMinBias2011Awf1000p0/all_OldVSNew_RunMinBias2011Awf1000p0c_recoJetedmRefToBaseProdTofloatsAssociationVector_ghostTrackBJetTags__RECO_obj_data_84.png

This PR was based on smth right before CMSSW_7_4_X_2015-01-23-0200
and jenkins tested on top of CMSSW_7_4_X_2015-01-23-0200, which is a correct choice.

Dr15Jones commented:
I only noticed a problem with crashes. For example
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc491/CMSSW_7_4_THREADED_X_2015-01-20-1400/pyRelValMatrixLogs/run/22.0_SingleMuPt1000+SingleMuPt1000INPUT+DIGI+RECO+HARVEST/step3_SingleMuPt1000+SingleMuPt1000INPUT+DIGI+RECO+HARVEST.log

I know of no validation on the threading which has been done. However, the problem I found could definitely also lead to incorrect results rather than a crash.

slava77 commented:
Would this lead to incorrect results only in MT or also in the single-thread?

Dr15Jones commented:
Only MT

Dr15Jones commented:
@slava77 How do you want to proceed?

slava77 commented:
me or Carl will need to run tests manually to confirm that it's not an issue with jenkins tests.
If the change is confirmed, we may need to go to the developers.

slava77 commented:
.. i will run the tests

Dr15Jones commented:
I'm also trying to see if I can create real unit tests for this code.

slava77 commented:
we have a bunch of duplicates like the following in the IB

```
LCGReflex/std::vector<reco::CATopJetTagInfo>
   **pluginDataFormatsBTauRecoCapabilities.so**
   **pluginDataFormatsJetRecoCapabilities.so**

```

This PR with dependencies needs recompilation of  packages using both of the above plugin*.so.
could this somehow trigger the effect?


@ktf 
Giulio, I think we will need a clean build to get rid of the duplicate definitions

slava77 commented:
the duplicates are a result of a move done in #7298
Apparently one of the two in the list is still from the original full-build IB, while the other one is from the patch-build

Dr15Jones commented:
To properly build, I had to 
1) rebase onto CMSSW_7_4_X
2) checkout DataFormats/JetReco in addition to this change

the reason was classes were moved since the IB I was using to test and what was in the head of CMSSW_7_4_X.

Dr15Jones commented:
I believe this code needs to be tested in an IB containing #7298

slava77 commented:
#7298 is in CMSSW_7_4_X_2015-01-22-0200

the discussion of the jenkins tests and the duplicates is for CMSSW_7_4_X_2015-01-23-0200

slava77 commented:
I was able to reproduce the discrepancies in CMSSW_7_4_X_2015-01-23-0200.
I'd rather wait for a full IB.
... staring at the showIB for a bit longer:

slc6_amd64_gcc481 CMSSW_7_4_X_2015-01-22-0200 is a full build, I was using gcc491, which turns out to be a patch.
I can give this one a try to sign this PR

Dr15Jones commented:
I think that my changes could change the behavior if a ProcForeach was a 'child' of another ProcForeach. However, if I'm reading the code correctly, the old behavior was also wrong since the 'child' ProcForeach would never have advanced and only ever returned its first value. 
I'm trying to build a unit test but so far trying to configure a MVAComputer by hand is proving to be neigh impossible.

slava77 commented:
I see, so, it sounds like it could be a real thing and my incomplete build tests maybe just after a red herring

slava77 commented:
@ferencek @arizzi
please take a look

Dr15Jones commented:
@slava77 I was able to make a stand alone unit test. It does show that this implementation has a bug.

Dr15Jones committed:
Added unit test for MVAComputer

Dr15Jones commented:
I've added the test to the pull request.

cmsbuild commented:
Pull request #7332 was updated. @cmsbuild, @vadler, @nclopezo, @monttj can you please check and sign again.

Dr15Jones committed:
bug fix: ProcForeach check condition needed a 'not'

When initializing the value of 'size' for the LoopCtx we need to
check that the 'size' has not yet been set. The bug only changed
the value of 'size' if it had already been set.

Dr15Jones commented:
@slava77 this last commit fixed the bug demonstrated by my unit test. Hopefully now everything should compare identically. Thanks for catching the problem.

Dr15Jones commented:
please test

cmsbuild commented:
Pull request #7332 was updated. @cmsbuild, @vadler, @nclopezo, @monttj can you please check and sign again.

slava77 commented:
@cmsbuild please test

maybe the bot listens to me

slava77 commented:
nope, it didn't listen to me either :(

cmsbuild commented:
+1
Tested at: f5ce02f7db5d67420ebe10f160acebc3b2da086e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-7332/2090/summary.html

slava77 commented:
Hi Chris,
your last fix did it. Now the comparisons are clean.

Dr15Jones commented:
Woo hoo! Thanks for doing all the checking.

ktf commented:
Bypassing analysis signature. @monttj complain if not ok.
